central/reports/service/v2/service_impl.go

```diff
@@ -191,3 +196,60 @@ func (s *serviceImpl) CancelReport(ctx context.Context, req *apiV2.ResourceByID)
 
 	return &apiV2.Empty{}, nil
 }
+
+func (s *serviceImpl) DownloadReport(ctx context.Context, req *apiV2.DownloadReportRequest) (*apiV2.DownloadReportResponse, error) {
+	if req == nil || req.GetId() == "" {
+		return nil, errors.Wrap(errox.InvalidArgs, "Empty request or report job id")
+	}
+
+	slimUser := authn.UserFromContext(ctx)
+	if slimUser == nil {
+		return nil, errors.New("Could not determine user identity from provided context")
+	}
+
+	rep, found, err := s.snapshotDatastore.Get(ctx, req.GetId())
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error finding report snapshot with job ID %q.", req.GetId())
+	}
+
+	if !found {
+		return nil, errors.Wrapf(errox.NotFound, "Error finding report snapshot with job ID '%q'.", req.GetId())
+	}
+
+	if slimUser.GetId() != rep.GetRequester().GetId() {
+		return nil, errors.Wrap(errox.NotAuthorized, "Report cannot be downloaded by a user who did not request the report.")
+	}
+
+	status := rep.GetReportStatus()
+	if status.GetReportNotificationMethod() != storage.ReportStatus_DOWNLOAD {
+		return nil, errors.Wrapf(errox.InvalidArgs, "Report job id %q did not generate a downloadable report and hence report cannot be downloaded.", req.GetId())
+	}
+
+	if status.GetRunState() == storage.ReportStatus_FAILURE {
+		return nil, errors.Errorf("Report job %q has failed and hence no report to download", req.GetId())
+	}
+	if status.GetRunState() != storage.ReportStatus_SUCCESS {
+		return nil, errors.Errorf("Report job %q is not ready for download", req.GetId())
+	}
+
+	// Fetch data
+	buf := bytes.NewBuffer(nil)
+
+	ctx = sac.WithGlobalAccessScopeChecker(ctx,
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+			sac.ResourceScopeKeys(resources.Administration)),
+	)
+
+	_, exists, err := s.blobStore.Get(ctx, common.GetReportBlobPath(req.GetId(), rep.GetReportConfigurationId()), buf)
+	if err != nil {
+		return nil, errors.Wrap(errox.InvariantViolation, "Failed to fetch report data")
+	}
+
+	if !exists {
+		// If the blob does not exist, report error.
+		return nil, errors.Errorf("Report is not available to download for job %q", req.GetId())
```

Comment:
`errors.Wrap(error.NotFound, ...)` here because snapshot can last longer (for history) than the blob.